From ff9a4546cbfbd8bee9dc44527a19368f3f7b3b1e Mon Sep 17 00:00:00 2001
From: Michal Srb <michalsrb@gmail.com>
Date: Sat, 18 Feb 2017 22:10:07 +0200
Subject: [PATCH 01/13] Refactor FcCompareValueList to take value_strong and
 value_weak.

Instead of complete value array. This will allow easier rewrite of FcFontMatch.
---
 src/fcmatch.c | 33 +++++++++++++++++++--------------
 1 file changed, 19 insertions(+), 14 deletions(-)

diff --git a/src/fcmatch.c b/src/fcmatch.c
index 40efbd3..731329a 100644
--- a/src/fcmatch.c
+++ b/src/fcmatch.c
@@ -357,7 +357,8 @@ FcCompareValueList (FcObject	     object,
 		    FcValueListPtr   v1orig,	/* pattern */
 		    FcValueListPtr   v2orig,	/* target */
 		    FcValue         *bestValue,
-		    double          *value,
+		    double          *value_strong,
+		    double          *value_weak,
 		    int             *n,
 		    FcResult        *result)
 {
@@ -417,18 +418,18 @@ FcCompareValueList (FcObject	     object,
 	FcValueListPrint (v2orig);
 	printf ("\n");
     }
-    if (value)
-    {
-	int weak    = match->weak;
-	int strong  = match->strong;
-	if (weak == strong)
-	    value[strong] += best;
+
+    if (value_strong) {
+	if (value_strong == value_weak)
+	    *value_strong += best;
 	else
-	{
-	    value[weak] += bestWeak;
-	    value[strong] += bestStrong;
-	}
+	    *value_strong += bestStrong;
     }
+    if (value_weak) {
+	if (value_strong != value_weak)
+	    *value_weak += bestWeak;
+    }
+
     if (n)
 	*n = pos;
 
@@ -466,10 +467,14 @@ FcCompare (FcPattern	*pat,
 	else
 	{
 	    const FcMatcher *match = FcObjectToMatcher (elt_i1->object, FcFalse);
+	    double *value_strong = (match ? &value[match->strong] : NULL);
+	    double *value_weak = (match ? &value[match->weak] : NULL);
+
 	    if (!FcCompareValueList (elt_i1->object, match,
 				     FcPatternEltValues(elt_i1),
 				     FcPatternEltValues(elt_i2),
-				     NULL, value, NULL, result))
+				     NULL, value_strong, value_weak,
+				     NULL, result))
 		return FcFalse;
 	    i1++;
 	    i2++;
@@ -531,7 +536,7 @@ FcFontRenderPrepare (FcConfig	    *config,
 
 		if (!FcCompareValueList (pel->object, match,
 					 FcPatternEltValues (pel),
-					 FcPatternEltValues (fel), NULL, NULL, &n, &result))
+					 FcPatternEltValues (fel), NULL, NULL, NULL, &n, &result))
 		{
 		    FcPatternDestroy (new);
 		    return NULL;
@@ -590,7 +595,7 @@ FcFontRenderPrepare (FcConfig	    *config,
 	    const FcMatcher *match = FcObjectToMatcher (pe->object, FcFalse);
 	    if (!FcCompareValueList (pe->object, match,
 				     FcPatternEltValues(pe),
-				     FcPatternEltValues(fe), &v, NULL, NULL, &result))
+				     FcPatternEltValues(fe), &v, NULL, NULL, NULL, &result))
 	    {
 		FcPatternDestroy (new);
 		return NULL;
-- 
2.15.1

